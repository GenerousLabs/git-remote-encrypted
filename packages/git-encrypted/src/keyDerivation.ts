import { scrypt } from 'scrypt-js';

export const hash = async (password: string, salt: string) => {
  const passwordBuffer = Buffer.from(password.normalize('NFKC'));
  const saltBuffer = Buffer.from(salt.normalize('NFKC'));

  const N = 1024;
  const r = 8;
  const p = 1;
  const dkLen = 96;

  const hash = await scrypt(passwordBuffer, saltBuffer, N, r, p, dkLen);

  return hash;
};
