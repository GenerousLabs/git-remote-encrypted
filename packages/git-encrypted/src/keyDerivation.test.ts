import { hash } from './keyDerivation';

describe('keyDerivation', () => {
  describe('hash()', () => {
    it('Creates a hash from the password "foobar"', () => {
      return expect(hash('foobar', 'foosalt')).resolves.toMatchSnapshot();
    });

    it('Returns a hash with a length of 96 bytes', () => {
      return expect(hash('quxbaz', 'deadbeef')).resolves.toHaveLength(96);
    });
  });
});
