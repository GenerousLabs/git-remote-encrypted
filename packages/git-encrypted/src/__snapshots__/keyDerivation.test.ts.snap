// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyDerivation hash() Creates a hash from the password "foobar" 1`] = `
Uint8Array [
  53,
  105,
  10,
  98,
  120,
  129,
  33,
  209,
  213,
  106,
  77,
  71,
  183,
  192,
  25,
  169,
  130,
  118,
  113,
  6,
  38,
  63,
  211,
  119,
  177,
  84,
  41,
  79,
  123,
  115,
  236,
  77,
  30,
  27,
  86,
  80,
  168,
  252,
  242,
  231,
  159,
  5,
  178,
  93,
  68,
  184,
  251,
  232,
  35,
  55,
  191,
  217,
  177,
  255,
  80,
  185,
  13,
  93,
  107,
  48,
  19,
  62,
  0,
  150,
  109,
  159,
  50,
  223,
  31,
  12,
  56,
  171,
  18,
  94,
  2,
  126,
  218,
  150,
  195,
  79,
  52,
  100,
  115,
  244,
  212,
  130,
  237,
  148,
  164,
  147,
  116,
  47,
  2,
  129,
  171,
  80,
]
`;
